substrate/frame/revive/src/benchmarking.rs

```diff
@@ -2330,6 +2336,245 @@ mod benchmarks {
+	/// - Pre-populates `InflightTransactions::<T>` storage with test data
+	///
+	/// Note: While trie root computation is theoretically O(N log N), the computational cost
+	/// is dominated by storage I/O operations, so linear regression approximation is adequate.
+	///
+	/// ## Usage:
+	/// - Fixed cost: `on_finalize(0, 0)` - cost incurred no matter what number of transactions
+	/// - Per transaction: `on_finalize(1, payload_size) - fixed_cost`
+	#[benchmark(pov_mode = Measured)]
+	fn on_finalize(n: Linear<0, 200>, p: Linear<0, 1000>) -> Result<(), BenchmarkError> {
+		let (instance, _origin, _storage_deposit, evm_value, signer_key, current_block) =
+			setup_finalize_block_benchmark::<T>()?;
+
+		// Initialize block
+		let _ = Pallet::<T>::on_initialize(current_block);
+
+		// Pre-populate InflightTransactions with n real transactions of input size p
+		// This isolates the on_finalize processing cost from transaction execution
+		if n > 0 {
+			// Create input data of size p for realistic transaction payloads
+			let input_data = vec![0x42u8; p as usize];
+			let gas_consumed = Weight::from_parts(1_000_000, 1000);
+
+			for _ in 0..n {
+				// Create real signed transaction with input data of size p
+				let signed_transaction = create_signed_transaction::<T>(
+					&signer_key,
+					instance.address,
+					evm_value,
+					input_data.clone(),
+				);
+
+				Pallet::<T>::store_transaction(signed_transaction, true, gas_consumed);
+			}
+		}
+
+		#[block]
+		{
+			// Measure only the finalization cost with n transactions of size p
+			let _ = Pallet::<T>::on_finalize(current_block);
+		}
+
+		// Verify block was properly finalized
+		assert!(Pallet::<T>::block_author().is_some());
+
+		// Verify transaction count
+		assert_eq!(Pallet::<T>::eth_block().transactions.len(), n as usize);
+
+		// Verify contract events count
+		let contract_events = get_contract_events::<T>();
+		assert_eq!(contract_events.len(), 0);
+
+		Ok(())
+	}
+
+	/// Benchmark the `on_finalize` per-transaction component costs.
+	///
+	/// This benchmark measures the computational cost of processing individual
+	/// transaction components (events and their data) within the finalization process.
+	/// Uses a single transaction to isolate component-specific costs.
+	///
+	/// ## Parameters:
+	/// - `e`: Number of events per transaction
+	/// - `d`: Total size of event data in bytes
+	///
+	/// ## Test Setup:
+	/// - Creates 1 transaction with specified parameters
+	/// - Transaction has `e` ContractEmitted events
+	/// - `d` total bytes of data across all events
+	///
+	/// ## Usage:
+	/// Calculate marginal costs:
+	/// - Per event: `on_finalize_per_transaction(1, 0) - on_finalize_per_transaction(0, 0)`
+	/// - Per byte of event data: `on_finalize_per_transaction(1, d) -
+	///   on_finalize_per_transaction(1, 0)`
+	#[benchmark(pov_mode = Measured)]
+	fn on_finalize_per_transaction(
+		e: Linear<0, 100>,
+		d: Linear<0, 16384>,
+	) -> Result<(), BenchmarkError> {
```

Comment:
I wonder if that work with the benchmark framework, or if we should create another benchmark as you initially suggested.

@ggwpez is this correct to create a benchmark with two variable here where `e` is the number of events and `d` the total bytes of data for distributed over all the events.

rather than doing one benchmark with `e` events and a fixed data and another one with 1 event and `d` bytes of data

Reply:
Done